File: python/japanese.py
import os
import random
import re
import sys
import time


def clean(str):
    return re.sub(' +', ' ', str)


class QuitException(Exception):
    pass


class Category:
    def __init__(self, name):
        self.name = name
        self.english = "english"
        self.japanese = "japanese"

    def __repr__(self):
        return self.name


class Word:
    def __init__(self, english, japanese):
        self.english = english
        self.japanese = japanese


# Examples:
# [this] apple is red [right?]
# is the apple red?
#
class SubjectIsAdjectiveCategory(Category):
    def __init__(self):
        super().__init__("SUBJECT is ADJECTIVE")
        self.subject = "unset"
        self.adjective = "unset"

        self.format = Word(
            "%s %s %s %s",
            "%s %s %s %s %s"
        )

    def populate(self, subject, adjective):
        self.subject = subject
        self.adjective = adjective

        is_or_isnot = [
            Word("IS", "です"),
            Word("IS NOT", "じゃないです"),
        ]
        prefix = [
            Word("THE", ""),
            Word("THIS (NEAR ME)", "この"),
            Word("THAT (NEAR YOU)", "その"),
            Word("THAT (NOT NEAR)", "あの"),
        ]

        v = is_or_isnot[random.randint(0, len(is_or_isnot) - 1)]
        p = prefix[random.randint(0, len(prefix) - 1)]

        self.japanese = clean(self.format.japanese % (
            p.japanese,
            self.subject.japanese,
            "は",
            self.adjective.japanese,
            v.japanese,
        ))
        self.english = clean(self.format.english % (
            p.english,
            self.subject.english,
            v.english,
            self.adjective.english,
        ))


class SubjectVerbAdverbCategory(Category):
    def __init__(self):
        super().__init__("SUBJECT VERB ADVERB")
        self.subject = "unset"
        self.verb = "unset"
        self.adverb = "unset"

    def populate(self, subject, adjective):
        self.subject = subject
        self.adjective = adjective

        self.japanese = f"TODO"
        self.english = f"TODO"


class Adjective():
    def __init__(self, english, japanese):
        self.japanese = japanese
        self.english = english


class Noun():
    def __init__(self, english, japanese):
        self.japanese = japanese
        self.english = english


categories = [
    SubjectIsAdjectiveCategory(),
    #SubjectVerbAdverbCategory(),
]


nouns = [
    Noun("APPLE", "りんご"),
]


adjectives = [
    Adjective("RED", "あかい"),
]


def continue_or_quit(prompt=""):
    choice = input(prompt)
    if choice == "q":
        raise QuitException("quitting")


def main(argv):
    global categories

    if len(argv) > 1 and argv[1] != "all":
        names = argv[1].split(",")
        categories = list(filter(lambda c: c.name in names, categories))
        assert categories
    else:
        category = input(f"Category {categories}: ")
        if category != "":
            names = category.split(",")
            categories = list(filter(lambda c: c.name in names, categories))
        assert categories

    while True:
        category = categories[random.randint(0, len(categories) - 1)]
        if isinstance(category, SubjectIsAdjectiveCategory):
            print(category)
            category.populate(
                nouns[random.randint(0, len(nouns) - 1)],
                adjectives[random.randint(0, len(adjectives) - 1)],
            )
            print(category.english)
            continue_or_quit()
            print(category.japanese)
            continue_or_quit()
        elif isinstance(category, SubjectVerbAdverbCategory):
            print(category)
        else:
            print("unknown category")

        continue_or_quit()


if __name__ == '__main__':
    try:
        main(sys.argv)
    except QuitException:
        pass

File: golang/japanese.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"time"
)

type Word struct {
	english  string
	japanese string
}

type Category interface {
	Name() string
	English() string
	Japanese() string
	Populate(args ...Word) bool
}

type SubjectIsAdjectiveCategory struct {
	english  string
	japanese string

	subject   Word
	adjective Word
}

func (c *SubjectIsAdjectiveCategory) Name() string {
	return "SUBJECT IS ADJECTIVE"
}

func (c *SubjectIsAdjectiveCategory) English() string {
	return c.english
}

func (c *SubjectIsAdjectiveCategory) Japanese() string {
	return c.japanese
}

func (c *SubjectIsAdjectiveCategory) Populate(args ...Word) bool {
	space := regexp.MustCompile(`\s+`)
	format := Word{
		"%s %s %s %s",
		"%s %s %s %s %s",
	}

	is_or_isnot := []Word{
		{"IS", "です"},
		{"IS NOT", "じゃないです"},
		{"IS ALSO", "です"},
		{"IS ALSO NOT", "じゃないです"},
	}
	prefix := []Word{
		{"THE", ""},
		{"THIS (NEAR ME)", "この"},
		{"THAT (NEAR YOU)", "その"},
		{"THAT (NOT NEAR)", "あの"},
	}

	c.subject = args[0]
	c.adjective = args[1]

	v := is_or_isnot[rand.Intn(len(is_or_isnot)-1)]
	p := prefix[rand.Intn(len(prefix)-1)]
	j := "は"
	if strings.Contains(v.english, "ALSO") {
		j = "も"
	}

	c.japanese = space.ReplaceAllString(fmt.Sprintf(format.japanese,
		p.japanese,
		c.subject.japanese,
		j,
		c.adjective.japanese,
		v.japanese,
	), " ")

	c.english = space.ReplaceAllString(fmt.Sprintf(format.english,
		p.english,
		c.subject.english,
		v.english,
		c.adjective.english,
	), " ")

	return true
}

type AnotherCategory struct {
	english  string
	japanese string

	something Word
}

func (c *AnotherCategory) Name() string {
	return "SOMETHING"
}

func (c *AnotherCategory) English() string {
	return c.english
}

func (c *AnotherCategory) Japanese() string {
	return c.japanese
}

func (c *AnotherCategory) Populate(args ...Word) bool {
	c.something = args[0]

	return true
}

var Nouns = []Word{
	{"APPLE", "りんご"},
}

var Adjectives = []Word{
	{"RED", "あかい"},
}

func Japanese() string {
	// fmt.Fprintf(os.Stderr, "japanese %s", os.Args[0])

	var categories = []Category{
		&SubjectIsAdjectiveCategory{
			english: "english", japanese: "japanese",
		},
		&AnotherCategory{
			english: "english something", japanese: "japanese something",
		},
	}

	reader := bufio.NewReader(os.Stdin)
	for {
		clear()
		categories[0].Populate(Nouns[0], Adjectives[0])

		fmt.Fprintf(os.Stderr, "%s\n", categories[0].English())
		if text, _ := reader.ReadString('\n'); text == "q\n" {
			break
		}
		fmt.Fprintf(os.Stderr, "%s\n", categories[0].Japanese())
		if text, _ := reader.ReadString('\n'); text == "q\n" {
			break
		}
	}

	return "japanese"
}

func clear() {
	cmd := exec.Command("clear") //Linux example, its tested
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func main() {
	rand.Seed(time.Now().Unix())
	fmt.Println(Japanese())
}
